File: rust/src/bin/stream_shm.rs
use libc::{close, mmap, munmap, shm_open};
use libc::{MAP_FAILED, MAP_SHARED, O_RDWR, PROT_READ, PROT_WRITE};
use std::ffi::CString;
use std::mem::size_of;
use std::ptr;
use std::thread::sleep;
use std::time::Duration;

// 常量定義
const SHM_NAME: &str = "/msg_queue";
const MAX_FLOAT_SIZE: usize = 16;
const MAX_QUEUE_SIZE: usize = 100000;

// 消息結構體定義
#[repr(C)]
// #[derive(Debug, Copy, Clone)]
struct Msg {
    // 1: Binance-Futures_BTCUSDT, 2: Binance-Futures_ETHUSDT, 3: Binance-Futures_SOLUSDT, 4: Binance-Futures_DOGEUSDT, 5: Binance_BTCUSDT
    instrument_id: i64,
    // 1: L1 Bid, -1: L1 Ask, 2: L2 Bid, -2: L2 Ask, 3: Buy Trade, -3: Sell Trade
    msg_type: i64,
    // Transaction Time MS
    tx_ms: i64,
    // Event Time MS
    event_ms: i64,
    // Local Time NS
    local_ns: i64,
    // Sequence Number / Trade ID
    sn_id: i64,
    // Price
    price: [u8; MAX_FLOAT_SIZE],
    // Size
    size: [u8; MAX_FLOAT_SIZE],
}

#[repr(C)]
// #[derive(Debug, Copy, Clone)]
struct Queue {
    sn: i64,
    from: usize,
    to: usize,
    msgs: [Msg; MAX_QUEUE_SIZE],
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // 打開共享記憶體區域
    let shm_name = CString::new(SHM_NAME)?;
    let fd = unsafe { shm_open(shm_name.as_ptr(), O_RDWR, 0o666) };
    if fd == -1 {
        return Err("Failed to open shared memory".into());
    }

    println!("Shared memory opened successfully");

    // 映射共享記憶體
    let ptr = unsafe {
        mmap(
            ptr::null_mut(),
            size_of::<Queue>(),
            PROT_READ | PROT_WRITE,
            MAP_SHARED,
            fd,
            0,
        )
    };

    if ptr == MAP_FAILED {
        unsafe { close(fd) };
        return Err("Failed to map shared memory".into());
    }

    println!("Memory mapped successfully");

    // 讀取共享記憶體中的 Queue 結構體數據
    let buf = ptr as *mut Queue;
    let mut cur;

    unsafe {
        cur = (*buf).from;
    }

    // 循環讀取消息
    loop {
        unsafe {
            while cur != (*buf).to {
                // 計算消息位置
                let msg = &(*buf).msgs[cur];

                // 將字符數組轉換為字符串
                let price_str = String::from_utf8_lossy(&msg.price);
                let size_str = String::from_utf8_lossy(&msg.size);

                // 打印消息
                println!(
                    "{}: {}, {}, {}, {}, {}, {}, {}, {}",
                    cur,
                    msg.instrument_id,
                    msg.msg_type,
                    msg.sn_id,
                    msg.tx_ms,
                    msg.event_ms,
                    msg.local_ns,
                    price_str,
                    size_str
                );

                // 更新指針
                cur = (cur + 1) % MAX_QUEUE_SIZE;
            }
        }

        // 短暫休眠以減少CPU使用率
        sleep(Duration::from_micros(1));
    }

    // 注意：這段代碼實際上永遠不會執行到，因為上面的循環是無限的
    // 但為了代碼完整性，保留以下清理代碼
    // unsafe {
    //     munmap(ptr, size_of::<Queue>());
    //     close(fd);
    // }

    Ok(())
}

File: python/stream_shm.py
import mmap
import os
import struct

# 定義常量
SHM_NAME = "/dev/shm/msg_queue"
MAX_FLOAT_SIZE = 16
MAX_QUEUE_SIZE = 100000
MSG_SIZE = 80  # 每個 Msg 的大小
SHM_SIZE = 8 + 8 + 8 + (MAX_QUEUE_SIZE * MSG_SIZE)  # Queue 結構的大小

# 定義 Msg 結構的格式 (與 C 的 struct 對應)
MSG_FORMAT = f"qqqqqq{MAX_FLOAT_SIZE}s{MAX_FLOAT_SIZE}s"

# Queue 偏移量和格式
QUEUE_HEADER_SIZE = 8 + 8 + 8  # sn, from, to 三個 long 的大小

# 開啟共享記憶體
fd = os.open(SHM_NAME, os.O_RDWR)
shm = mmap.mmap(fd, SHM_SIZE, mmap.MAP_SHARED, mmap.PROT_READ | mmap.PROT_WRITE)

# 讀取 Queue 結構
def read_queue_header():
    shm.seek(0)
    header_data = shm.read(QUEUE_HEADER_SIZE)
    sn, from_index, to_index = struct.unpack("qqq", header_data)
    return sn, from_index, to_index

# 主循環
def main():
    _, from_index, to_index = read_queue_header()
    cur = from_index

    while True:
        # 更新 Queue Header
        _, from_index, to_index = read_queue_header()

        while cur != to_index:
            msg_offset = QUEUE_HEADER_SIZE + (cur * MSG_SIZE)
            shm.seek(msg_offset)
            raw_msg = shm.read(MSG_SIZE)

            # 確保有足夠的數據
            if len(raw_msg) < MSG_SIZE:
                raise ValueError(f"Buffer size too small ({len(raw_msg)} instead of at least {MSG_SIZE} bytes)")

            # 解構 msg
            (
                # 1: Binance-Futures_BTCUSDT, 2: Binance-Futures_ETHUSDT, 3: Binance-Futures_SOLUSDT, 4: Binance-Futures_DOGEUSDT, 5: Binance_BTCUSDT
                instrument_id,
                # 1: L1 Bid, -1: L1 Ask, 2: L2 Bid, -2: L2 Ask, 3: Buy Trade, -3: Sell Trade
                msg_type,
                # Transaction Time MS
                tx_ms,
                # Event Time MS
                event_ms,
                # Local Time NS
                local_ns,
                # Sequence Number / Trade ID
                sn_id,
                # Price
                price,
                # Size
                size
            ) = struct.unpack(MSG_FORMAT, raw_msg)

            # 打印消息
            print(
                f"{cur}: {instrument_id}, {msg_type}, {sn_id}, {tx_ms}, {event_ms}, {local_ns}, {price.decode().strip()}, {size.decode().strip()}"
            )

            # 更新指針
            cur = (cur + 1) % MAX_QUEUE_SIZE

if __name__ == "__main__":
    try:
        main()
    finally:
        shm.close()
        os.close(fd)
